# Flight Impact - Data Cleaning Notebook

This project focuses on creating a Streamlit web app using flight emissions data to help educate consumers about their carbon impact. In order to develop a fast-moving app, I need to build a dataset using numerous calls to the ClimatIQ API, and store the data in a MongoDB database. The app will then load data from the database when it is launched. 

First, I'll begin by creating a dataset of all combinations of the world's 'large' (serving millions per year) airports. This will be the foundation for my API calls and will allow for users to explore a wide variety of routes. 

Please note - the original version of this notebook was located in Google Colab. 

In [2]:
import requests
import json
import pandas as pd
import numpy as np
from pymongo import MongoClient
from itertools import combinations
from geopy.distance import great_circle

In [3]:
#dataset of the world's airport codes from https://datahub.io/core/airport-codes
airport_data = pd.read_csv('/content/drive/My Drive/airport-codes_csv.csv')
airport_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57421 entries, 0 to 57420
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ident         57421 non-null  object 
 1   type          57421 non-null  object 
 2   name          57421 non-null  object 
 3   elevation_ft  49608 non-null  float64
 4   continent     28978 non-null  object 
 5   iso_country   57175 non-null  object 
 6   iso_region    57421 non-null  object 
 7   municipality  51527 non-null  object 
 8   gps_code      41561 non-null  object 
 9   iata_code     9225 non-null   object 
 10  local_code    30030 non-null  object 
 11  coordinates   57421 non-null  object 
dtypes: float64(1), object(11)
memory usage: 5.3+ MB


In [4]:
airport_data = airport_data.dropna(axis=0, subset=['iata_code'])
airport_data = airport_data.drop(labels=['elevation_ft', 'local_code', 'gps_code'], axis=1)


In [5]:
airport_codes = airport_data.iata_code.unique()
#we just want to use large airports, since these are most frequented by travelers
#and our API call amount is limited
large_airport_codes = airport_data[(airport_data['type']=='large_airport')]


In [6]:
#flip the coords around and convert them to a tuple
def fix_coords(row):
    order = [1, 0]
    split_stripped_coords = [x.strip() for x in row.split(',')]
    reordered = tuple(split_stripped_coords[i] for i in order)
    return reordered

large_airport_codes['correct_coords'] = large_airport_codes.coordinates.apply(fix_coords)
large_airport_codes=large_airport_codes.drop(labels=['coordinates', 'type'],axis=1)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [7]:
large_airport_codes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 606 entries, 11927 to 57414
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ident           606 non-null    object
 1   name            606 non-null    object
 2   continent       402 non-null    object
 3   iso_country     605 non-null    object
 4   iso_region      606 non-null    object
 5   municipality    600 non-null    object
 6   iata_code       606 non-null    object
 7   correct_coords  606 non-null    object
dtypes: object(8)
memory usage: 42.6+ KB


There are a total of 606 large global airports in our dataset. Next, we want to find every possible combination of two airports, so our dataset will have all global routes. For storage purposes, we will not include every permutation such as (CHI, DEN) and (DEN, CHI) - a single combination of two airports will only occur once. 

In [8]:
combs = list(combinations(large_airport_codes.iata_code.unique(), 2))

In order to get the most accurate data back from the ClimatIQ API, we need to know the distance between each pair of airports. Below we'll calculate 'great circle' distance, which is the shortest distance between two points on a globe. Then we'll append all needed information to a list to create our robust datset for API calls. 

In [9]:
list_to_df = []

for combination in combs: 
    loc1_coords = large_airport_codes.correct_coords[large_airport_codes.iata_code==combination[0]].iloc[0]
    loc2_coords = large_airport_codes.correct_coords[large_airport_codes.iata_code==combination[1]].iloc[0]

    loc1_name = large_airport_codes.name[large_airport_codes.iata_code==combination[0]].iloc[0]
    loc2_name = large_airport_codes.name[large_airport_codes.iata_code==combination[1]].iloc[0]


    gc_distance = great_circle(loc1_coords, loc2_coords).mi 
    list_to_df.append([combination[0],loc1_name, loc1_coords, combination[1],loc2_name, loc2_coords, gc_distance])

In [11]:
list_to_df[:3]

[['POM',
  'Port Moresby Jacksons International Airport',
  ('-9.443380355834961', '147.22000122070312'),
  'KEF',
  'Keflavik International Airport',
  ('63.985001', '-22.6056'),
  8635.37883637017],
 ['POM',
  'Port Moresby Jacksons International Airport',
  ('-9.443380355834961', '147.22000122070312'),
  'PRN',
  'Priština International Airport',
  ('42.5728', '21.035801'),
  8476.052502788527],
 ['POM',
  'Port Moresby Jacksons International Airport',
  ('-9.443380355834961', '147.22000122070312'),
  'YEG',
  'Edmonton International Airport',
  ('53.309700012200004', '-113.580001831'),
  7120.067716968816]]

In [12]:
#turn our list into a dataframe
route_distances = pd.DataFrame(list_to_df, columns=['origin_code','origin_name', 'origin_coords', 'dest_code', 'dest_name', 'dest_coords', 'gc_distance'])

In [13]:
route_distances.head(3)

origin_code  ...  gc_distance
0         POM  ...  8635.378836
1         POM  ...  8476.052503
2         POM  ...  7120.067717

[3 rows x 7 columns]

To successfully automate our API calls, we need to get the parameters for each specific call, which are dependent on the route distance. The EPA's emissions calculations are divided up by short haul (<300 miles), medium haul (300-2300 miles), and long haul (>2300 miles) flights. Below we use the route distance to get the API call parameters for each route. 

In [14]:
def get_params(row):
    if row[6]<300:
        id = "passenger_flight-route_type_na-aircraft_type_na-distance_lt_300mi-class_na-contrails_na"
    elif (row[6] >=300) and (row[6]<2300):
        id = "passenger_flight-route_type_na-aircraft_type_na-distance_gt_300mi_lt_2300mi-class_na-contrails_na"
    else:
        id = "passenger_flight-route_type_na-aircraft_type_na-distance_gt_2300mi-class_na-contrails_na"

    route = [row[0], row[3]]

    return {"emission_factor":id, "parameters": {"route": route}, "persist":True}

route_distances["query_params"] = route_distances.apply(get_params, axis=1)

## Database Setup

MongoDB is the storage solution that we'll use for our large dataset. We need to set up a MongoDB database into which we can insert any data that is given to us from the API. 

In [19]:
import pymongo
from pymongo import MongoClient
from getpass import getpass

In [20]:
uri = 'mongodb://urhejh70922nhwipt6kt:hNfQXzFxrsDQGGbyH8KX@bs8ntk4apfl7fga-mongodb.services.clever-cloud.com:27017/bs8ntk4apfl7fga'
client = MongoClient( uri )

In [21]:
# MongoDB connection info
hostname = 'bs8ntk4apfl7fga-mongodb.services.clever-cloud.com'
port = 27017
username = 'urhejh70922nhwipt6kt'
password = getpass('Enter the secret value: ')
databaseName = 'bs8ntk4apfl7fga'

# connect with authentication
client = MongoClient(hostname, port)
db = client[databaseName]
db.authenticate(username, password)


Enter the secret value: ··········


True

In [22]:
#db.create_collection('final_flight_app')

## ClimatIQ API Calls

ClimatIQ API has some limits on how many API calls you can make in a second, so we built in some sleep time to not exceed their constraints. The for loop below will call the API for each row in the dataframe, using the params we found earlier to specify which calculations should be used to get the flight emissions. It will then return to us the CO2 emissions of that flight in kg for a single economy passenger, one way. 

In [ ]:
# import time

#results_dict = {}

# for idx, params in enumerate(route_distances.query_params[:10]):
#     payload = json.dumps(params)
# #   headers = {"Authorization": config.climatiq_bearer_key, "Content-Type": "application/json"}
# #   request_dict = requests.post('https://beta2.api.climatiq.io/estimate',data=payload, headers=headers).json()
#     request_dict = {'test':'test'}
#     info_dict = dict(route_distances.iloc[idx])
#     info_dict.pop('query_params')
#     to_insert = dict(info_dict, **request_dict)
  
# #   results_dict[idx] = to_insert
# #   time.sleep(0.1)

With this new collection of data, we'll do some cleaning & manipulation below that will give us more flexibility when creating our web app. At the end, we'll insert all of the data into our database. 

In [ ]:
flights = pd.DataFrame(results_dict)


In [ ]:
len(flights.dest_name.unique())

605

In [ ]:
flights = flights.dropna(subset = ['co2e'],axis=0).drop(labels=['error','message'], axis=1)


In [ ]:
#fixing some missing municipalities by googling them
large_airport_codes.loc[21154, 'municipality'] = 'Manzini'
large_airport_codes.loc[38859, 'municipality'] = 'Dhahran'
large_airport_codes.loc[41601, 'municipality'] = 'Pyeongtaek'
large_airport_codes.loc[41666, 'municipality'] = 'Okinawa'
large_airport_codes.loc[53056, 'municipality'] = 'Mattala'
large_airport_codes.loc[54346, 'municipality'] = 'Medan'

large_airport_codes.loc[22315, 'iso_country'] = 'NA'

large_airport_codes.continent[large_airport_codes.continent.isnull()] = 'NA'


In [ ]:
#merge emissions data with the airport info data including country & continent for the origin airport
flights = pd.merge(flights,large_airport_codes[['name', 'continent', 'iso_country', 'iso_region', 'municipality', 'iata_code']],how='inner',left_on='origin_code', right_on='iata_code')
flights.rename({'continent':'origin_continent', 'iso_country':'origin_country', 'iso_region':'origin_region', 'municipality':'origin_city'}, inplace=True,axis=1)

In [ ]:
flights.drop(labels=['iata_code'], axis=1,inplace=True)

In [ ]:
#join again, this time getting the info for the destination airport
flights = pd.merge(flights,large_airport_codes[['name', 'continent', 'iso_country', 'iso_region', 'municipality', 'iata_code']],how='inner',left_on='dest_code', right_on='iata_code')


In [ ]:
flights.rename({'continent':'dest_continent', 'iso_country':'dest_country', 'iso_region':'dest_region', 'municipality':'dest_city'}, inplace=True,axis=1)
flights.drop(labels=['iata_code','name_y', 'name_x'], axis=1,inplace=True)

In [ ]:
#when pulling from the API, special characters such as accents got messed up
#here we fix the city and country names that are misspelled with odd characters
replace_city_dict = {'PoznaÅ\x84':'Poznan', 'GdaÅ\x84sk':'Gdansk', 'WrocÅ\x82aw':'Wroclaw', 'Montréal':'Montreal', 'Bordeaux/Mérignac':'Bordeaux/Merignac',
                     'San José del Cabo': 'San Jose del Cabo', 'Belém':'Belem', 'Malé':'Male', 'Münster':'Munster','Düsseldorf':'Dusseldorf', 'Ã\x9crümqi':'Urumqi',
                     'Bâle/Mulhouse':'Basel/Mulhouse', 'São Paulo':'Sao Paulo', 'ReykjavÃ\xadk':'Reykjavik','Liège':'Liege', 'Cancún':'Cancun',
                     'Pasay / Parañaque, Metro Manila':'Pasay / Paranaque, Metro Manila','Florianópolis':'Florianopolis', 'Nápoli':'Na Poli', 'Hagåtña, Guam International Airport':'Hagatna',
                     'Tromsø':'Tromso','Kraków':'Krakow', 'Pointe-Ã\xa0-Pitre':'Point-a-Pitre', 'LuleÃ¥':'Lulea', 'Praia da Vitória':'Praia da Vitoria', 'Bodø':'Bodo', 'Málaga':'Malaga',
                     'BrasÃ\xadlia':'Brasilia','Malmö':'Malmo', 'Ä°zmir':'Izmir' }
for k, v in replace_city_dict.items():
    flights.origin_city[flights.origin_city == k] = v
    flights.dest_city[flights.dest_city == k] = v


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


In [ ]:
#these are new columns that will give us more flexibility when building our app
flights['origin_lat'] = flights.origin_coords.apply(lambda x:float(x[0]))
flights['origin_long'] = flights.origin_coords.apply(lambda x:float(x[1]))
flights['dest_lat'] = flights.dest_coords.apply(lambda x:float(x[0]))
flights['dest_long'] = flights.dest_coords.apply(lambda x:float(x[1]))
flights['formatted_co2e'] = flights['co2e'].apply(lambda x: int(x))
flights['formatted_tons'] = round(flights['formatted_co2e'] / 1000, 2)

In [ ]:
#iso codes will help us get the full names of the countries that our airports are located in
iso_codes = pd.read_csv('/content/drive/My Drive/country_iso_codes.csv')
iso_codes.Code[iso_codes.Name=='Namibia'] = 'NA'
iso_codes = iso_codes.append({'Name':'Kosovo', 'Code':'XK'}, ignore_index=True)

In [ ]:
#join iso codes for both origin and destination airports
flights = pd.merge(flights, iso_codes, left_on='origin_country', right_on='Code', how='left')
flights = pd.merge(flights, iso_codes, left_on='dest_country', right_on='Code', how='left')

In [ ]:
flights.head()

_id         co2e co2e_unit  ... Code_x   Name_y Code_y
0  61b26bb4203479d554e28590  1390.294028        kg  ...     PG  Iceland     IS
1  61b26bb4203479d554e28591  1364.642520        kg  ...     PG   Kosovo     XK
2  61b26d3c203479d554e287ed   295.769183        kg  ...     IS   Kosovo     XK
3  61b26bb4203479d554e28592  1146.329279        kg  ...     PG   Canada     CA
4  61b26d3c203479d554e287ee   492.355426        kg  ...     IS   Canada     CA

[5 rows x 33 columns]

In [ ]:
flights.rename({'Name_x':'origin_country_full', 'Name_y':'dest_country_full'},inplace=True,axis=1)
flights.drop(labels=['Code_x','Code_y'],axis=1,inplace=True)

In [ ]:
#the ISO code 'NA' for Namibia is reading as null 
flights.origin_country_full[flights.origin_country_full.isnull()] = 'Namibia'
flights.dest_country_full[flights.dest_country_full.isnull()] = 'Namibia'

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
flights_dict2 = flights.to_dict('records')

In [ ]:
#finally, insert our cleaned up dataset into the MongoDB database
db.final_flight_app.insert_many(flights_dict2)